[PRFC] Extendable components

grantila opened the pull request:
# [PRFC] Adaptable components

 * "Extendable" renamed "Adaptable"


## TL;DR

How to make an adaptable component:

```tsx
const { componentRef, MyComponent } =
  createAdaptableComponent<Props, Context>({
    Provider() { /* Prepare data, maybe async, provide context */ },
    Component() { /* The pure UI component, read above context */ },
  });
```

How to adapt a component:

```tsx
adaptComponent(componentRef, {
  Adaptation({ value, Component }) { 
    /* override default context values */
    return <Component value={ /* ... */ } />
  }
});
```


## Rationale

As raised in [#11035](https://github.com/backstage/backstage/issues/11035), there's a need to be able to configure (or _adapt_) components, especially core components. Implementors might want to add a tooltip around a component, intercept the `to` in Links (e.g to redirect internal corporate links `user:@joe`, `slack:#channel-name`), or add/change the UI of components (e.g. add custom action buttons to certain cards, enable/disable them given async logic).

If this is done globally at bootstrap time, it will affect such components being rendered _anywhere_. There will not be a need to customize this wherever components are _used_, which wouldn't be feasible.

It can also be useful to be able to scope this to a certain place in the React component tree, such that under a certain component, every `<Something>` is adapted a certain way. An example of this is the entity page header.

This can be done in several ways, but to be as powerful and useful as possible, it would be beneficial to be able to use asynchronicity (for async lookups), being able to handle errors, fallback to other UI's and maybe even tap into the React component lifecycle, so that the logic that adapts a component can e.g. subscribe to observables, and properly unsubscribe when the component is unloaded.

This PR provides a solution to this.


## Back to the future

Soon after React came about, the concept of Active and Passive components became popular. Sometimes called Smart/Dumb components, or Components/Containers.

The concept of this is to separate the _logic_ of components (networking, fetching, parsing, error handling, waiting) in Active components, to pure non-logic UI rendering of whatever is passed as props in Passive components.

Often, this was implemented using _render props_ which is ineffective as they are re-evaluated every render, and when React got Context and functional components (with hooks), logic was quickly moved into hooks instead.

The fundemantal idea of Active and Passive components can still be employed though, to enable adapting components in a pretty slick way.


## Provider/Component

For components to be adaptable, we need to _make them_ adaptable, unless we go down the dark route of _hacking_ into them.

To make a component adaptable, we can split it in two parts, but maintain its API surface to be just like before. Given the Active/Passive semantics described above, let's reason about them similarly under the names Provider and Component.

 1. [Provider] A component that takes the props and performs whatever logic necessary (could be async or anything else), and turns the props that should be configurable and any other parsed/fetched result, as a _context_ object to its inner Component.
 2. [Component] A component that reads its input from the context provided by the Provider. It also has access to the raw props, and can use that if necessary (could be values that shouldn't be configurable or just passing things like `className` downstream).

Example:

The link component has a prop `to`, but also practically any other HTML element prop, including `children`. The API would be the same:

```tsx
<Link to="url">Text</Link>
```

But the component is internally two components. We can choose to make `to` and `children` configurable by creating a `LinkContext` (side by side with `LinkProps`):

```ts
interface LinkContext {
  to: string;
  children: React.ReactNode;
}
```

In fact, in `LinkProps`, the `to` can be more than just a string, and handling this is something we can do in the LinkProvider. So when using a `<Link>`, this will turn into:

```tsx
<LinkProvider to="url">
  <LinkComponent />
</LinkProvider>
```

The `LinkComponent` will read the `to` and `children` from the context instead of from its props.

Now let's introduce _adaptations_ to the Link component. Adaptations are normal React components, and can intercept the context value. When adapting the Link component, the adaptations will become intermediate providers in between the Provider and Component:

```tsx
<LinkProvider to="url">
  <Adaptations> { /* <-- adaptations end up here where they can rewrite the context */ }
    <LinkComponent />
  </Adaptations>
</LinkProvider>
```


## Creating an adaptable component

To create an adaptable component, this PR proposes `createAdaptableComponent()` which returns the public API surfaced component, as well as a component _ref_ which can be used to _reference_ the component when adapting it.

Let's make this (very minimalistic) Link component adaptable:

```tsx
export type LinkProps = { /* ... */ };

export function Link(props: LinkProps) {
  const to = String(props.to); // Stringify if not string
  return <a className={props.className} href={to}>{props.children}</a>;
}
```

We do this using `createAdaptableComponent`, and by defining the `LinkContext` which is how the Provider communicates with the Component:

```tsx
export type LinkProps = { /* ... */ };
export type LinkContext = {
  to: string;
  children: React.ReactNode;
};

export const { componentRef: linkComponentRef, Component: Link } =
  createAdaptableComponent<LinkProps, LinkContext>({
    id: 'Link:v1',
    Provider: ({ props, Component }) => {
      const value: LinkContext = {
        to: String(props.to),
        children: props.children,
      };

      return <Component value={value} />;
    },
    Component: ({ props, value }) => {
      return <a className={props.className} href={value.to}>{value.children}</a>;
    },
  });
```

It might look like a lot of overhead, but remember this Link component is tiny. The extra work of splitting it in two is the same in absolute terms for more realistic components, and ends up being marginal.

We export `LinkContext`, even though it's not useful when _using_ the Link component (only the component itself and its `LinkProps` are). But the context type is important when _adapting_ the component, and even though likely not _necessary_ to use it explicitly, it can still be useful.

In the export destructuring we export `linkComponentRef` which can be thought of as an `apiRef` in that it's opaque and used for referencing, and `Link` (with the same API as the original `Link` component). This is not a breaking change, the public API only has additions.

By providing the Props and Context types to `createAdaptableComponent`, the two components (the provider and the "UI component") have typesafe props. Let's examine them:


### The Provider

The Provider part of the adaptable component is a component itself (hence can e.g. use hooks). Its props are:
 * `props` the props provided wherever `<Link>` is used in our example.
 * `Component` an opaque component. This will perhaps be the inner component, but it could also be intermediate component _adaptations_. It has a prop `value` (of type `LinkContext`) which needs to be set to the context value.


### The Component

The Component part of the adaptable component is what renders the component. If making an existing component adaptable, most (if not all) of its internal React output (the "HTML") ends up here. Its props are:
 * `props` the raw props object provided to the `<Link>` in this example.
 * `value` the context object provided by the Provider, and potentially adapted (altered) by a component adaptation.

### Lazy-loading

Both the `Provider` component, and the `Component` component can be lazy-loaded, by instead providing `asyncProvider` and `asyncComponent`:

```tsx
export const { componentRef: linkComponentRef, Component: Link } =
  createAdaptableComponent<LinkProps, LinkContext>({
    asyncProvider: () => import('./my-component').then(m => m.Provider),
    asyncComponent: () => import('./my-component').then(m => m.Component),
  });
```


## Adapting

If we want to adapt a component, we use `adaptComponent`. This can be used in the top-level app when setting it up, but plugins can also adapt components automatically as part of `createPlugin`. When plugins provide adaptations, they provide them in the `createPlugin` call, using `createComponentAdaptationExtension` rather than `adaptComponent`, to tie the adaptation to the plugin.

We can also locally adapt components somewhere in the React component tree (for sub components).

When adapting, not only can the context be rewritten, but what children components are rendered can be altered too. A component adaptation is a component, so it can return anything. Sometimes changing the context is useful, sometimes just changing what's _rendered_ is useful.

Adaptations are created using `adaptComponent()` (or if locally in a React component and depending on local component state, `useAdaptComponent()`). The following example is silly by logic, but shows the syntax:

```tsx
const adaptation = adaptComponent(linkComponentRef, {
  Adaptation({ value, Component }) {
    if (!canAdaptThis(value))
      // Fallback - can't adapt so just render the real component
      return <Component />;

    // Overwrite the context "to" field
    return <Component value={{ ...value, to: '/' }} />;
  },
});
```

The adaptation is a component itself. It can e.g. use hooks. It gets a mix of the props that the component itself gets:
 * `value` the context value (so far in the context stack)
 * `Component` an opaque component which renders the real inner component (or another intermediate adaptation). It has an optional prop `value` (like the Provider-side of the adaptable component) which can be used to set a new context value.
 * `props` the raw props (readonly), sometimes useful to make decisions from

The return value `adaptation` is opaque, and is either provided in `createApp`, in `createPlugin` or in `<ProvideAdaptation>` which enables adaptations scoped to a certain part of the React component tree. More on that below.


### Example: change what's being rendered (not the context)

Let's say the Github plugin should adapt the Link component to add badges after links that goes to PR's so that we see a colorful dot after the link for whether it's open, merged or closed. The `linkComponentRef` can be imported (just like an API ref), and when used with `adaptComponent` it automatically provides types to ensure the adaptation is typesafe with regards to the context it can alter.

```tsx
createPlugin({
  id: 'github',
  // ... ,
  adaptations: {
    link: createComponentAdaptationExtension(linkComponentRef, {
      id: 'github-link-adaptation',
      Adaptation({ value, Component }) {
        if (!isLinkToPr(value.to))
          // Fallback - can't adapt so just render the real component
          return <Component />;

        // Let's assume useGithubPrStatus exists and subscribes to some observable internally
        const status = useGithubPrStatus(getPrFromLink(value.to));

        return (
          !status // Maybe undefined on first rendering before observable emits
          ? <Component />
          : <Badge color={prColorFromStatus(status)}><Component /></Badge>
        );
      },
    }),
  }
});
```

The above clearly uses functions we don't have (`isLinkToPr`, `useGithubPrStatus`, `getPrFromLink` and `prColorFromStatus`), but, would be quite easy to write. If so, anywhere we render a regular `<Link>`, it could be decorated with PR/issue status badge. The same could happen for links to CI/CD builds, or anything else where it'd be useful to see status/progress live or just extra information.

If the Link component becomes adaptable, and it is being used in many different contexts (some of where UI decoration isn't always a good idea) perhaps adding a prop `noDecoration?: boolean` to it would make sense. The adaptations could then respect this, and not decorate it (but perhaps still change the `to` if that'd be useful).

How adaptable components work and when what kind of adaptations is a good idea, will likely be different for different components. The link is very generic, and used everywhere, other components are more esoteric.


### Example: change context, not chrome

Note that we didn't set a `value` for `<Component />` in the Github PR decorator adaptation, but rather wrapped the link with a `<Badge>` wrapper.

Other components could be adaptable in different ways. For example, a user Card showing the avatar and name of a user (perhaps in the Catalog), could have a set of action buttons at the bottom. If this card is made adaptable, the avatar, name and action buttons could all be turned into context values, adaptable/changeable by adaptations.

We can imagine such a card being used like this:

```tsx
<UserCard entityRef={ /* ... */ } />
```

and it could be implemented, as an adaptable component, like this:

```tsx
export type UserCardProps = { entityRef: string; };
export type UserCardContext = {
  avatarUrl: string;
  name: string;
  actionButtons: { href?: string; onClick?: () => void; text: string; }[];
};

export const { componentRef: userCardComponentRef, Component: UserCard } =
  createAdaptableComponent<UserCardProps, UserCardContext>({
    Adaptation: ({ props, Component }) => {
      const user = useEntityRef(props.entityRef);

      if (!user.value)
        return <Progress />;

      const value: UserCardContext = parseUser(user.value);

      return <Component value={value} />;
    },
    Component: ({ props, value }) => {
      return (
        <Card>
          <CardHeader>...</CardHeader>
          <CardContent>...</CardContent>
          <ActionButtons>...</ActionButtons>
        </Card>
      );
    },
  });
```

An adaptation could e.g. easily add its own action button:

```tsx
adaptComponent(userCardComponentRef, {
  Adaptation({ value, Component }) {
    if (!shouldHaveExtraActionButton(value))
      return <Component />;

    const newValue = {
      ...value,
      actionButtons: [
        ...value.actionButtons,
        {
          text: 'My action',
          onClick: () => { /* ... */ },
        }
      ]
    };

    return <Component value={newValue} />;
  },
}),
```


## Local adaptations in the React component tree

To add an adaptation to a component at a certain place in the component tree, use `ProvideAdaptation`. It has a prop `adaptations` which can be either 1 adaptation, or an array of adaptations.

`useAdaptComponent` is the same as `adaptComponent` but with a last _deps_ argument like most React hooks. It memoizes the adaptation as long as the _deps_ don't change.

```tsx
function MyComponent(props: MyProps) {
  // This adaptation depends on props.someProp
  const adaptation = useAdaptComponent(
    userCardComponentRef,
    {
      Adaptation({ value, Component }) {
        if (someCondition(props.someProp))
          return <Component />;

        const newValue = {
          ...value,
          ...myOverrides,
        };

        return <Component value={newValue} />;
      },
    },
    [props.someProp]
  );

  // The adaptation will only be used for UserCard components under
  // this part of the React component tree.
  return (
    <ProvideAdaptation adaptations={adaptation}>
      <MyOtherComponentsGoesHere />
    </ProvideAdaptation>
  );
}
```

## Lazy-loaded adaptations

Instead of providing the `Adaptation` component in `adaptComponent`, `useAdaptComponent` or `createComponentAdaptationExtension`, the implementation can be lazy-loaded and exposed in `asyncAdaptation`:

```tsx
const adaptation = adaptComponent(
  userCardComponentRef,
  {
    Adaptation({ value, Component }) {
      // ...
    },

    // or

    asyncAdaptation: () => import('./my-adaptation').then(m => m.Adaptation),
  },
);
```

This will lazy-load the adaptation. The target component (UserCard in this case) will be immediately rendered _without_ this adaptation (to not _lag_ the UI). When the adaptation has loaded, it will be _applied_.


## Props interception

Sometimes it makes sense to intercept props and alter them. This is not the same as injecting logic in-between the Provider and Component as above, but rather customizing props specific to the installation.

The `adaptComponent` and `useAdaptComponent` functions as used above, take an `Adaptation` (or `asyncAdaptation`) property, which is optional. The other optional property is `interceptProps`:

```ts
export type AdaptableComponentSpec = {
  /**
   * Intercept the props, to e.g. rewrite some of them
   */
  interceptProps?: (props) => typeof props;

  /**
   * An adaptation component to intercept the context
   */
  Adaptation?: /* */;
  asyncAdaptation?: () => Promise<typeof Adaptation>;
};
```

This is useful when e.g. rewriting the Link `to` to understand installation specific "magic" links, or perhaps to change props when running in dev-mode, during testing etc.

```ts
adaptComponent(
  linkComponentRef,
  {
    interceptProps: props => ({
      ...props,
      `${props.to}`.beginsWith('slack:#')
        ? fixSlackUrl(props.to)
        : props.to,
    }),
  }
);
```

When adaptations are applied, first all `interceptProps` functions get invoked, and then all `Adaptation` components get injected.


## Ref

`createAdaptableComponent()` has a sibling function `createAdaptableForwardableComponent()` in which case the _inner_ Component gets a React `ref` is such is provided. The resulting component (to be used anywhere) will then have a `ref` prop too.

## Separate component ref factory

It is sometimes useful to be able to create component refs separately from its implemenation. In the example above, we use `createAdaptableComponent` which is a shorthand for `createAdaptableComponentRef` followed by `implementAdaptableComponent`.

```tsx
export const { componentRef: linkComponentRef, Component: Link } =
  createAdaptableComponent<LinkProps, LinkContext>({
    id: 'link',
    Provider() { /* ... */ },
    Component() { /* ... */ },
  });

// is the same as:

export const linkComponentRef =
  createAdaptableComponentRef<LinkProps, LinkContext>({ id: 'link' });

export const Link =
  implementAdaptableComponent(linkComponentRef, {
    Provider() { /* ... */ },
    Component() { /* ... */ },
  });
```


## Wiring up the app

Plugins expose adaptations they provide using `createComponentAdaptationExtension`. The resulting plugin object will have an `adaptations` property with a typed object keyed to the provided adaptations:

```tsx
const githubPlugin = createPlugin({
  id: 'github',
  // ... ,
  adaptations: {
    link: createComponentAdaptationExtension(linkComponentRef, {
      id: 'github-link-adaptation',
      Adaptation({ value, Component }) { /* ... */ },
    }),
  }
});

githubPlugin.adaptations.link; // is the adaptation above
```

When wiring up the app using `createApp`, we can choose which adaptations to use:

```tsx
createApp({
  // ...
  adaptations: [
    githubPlugin.adaptations.link,
    pluginX.adaptations.y,
    // ...
  ],
});
```

We can also choose to use all adaptations provided by the plugins by default, and opt-out of certain adaptations (or entire plugins), as well as including non-plugin provided adaptations:

```tsx
createApp({
  // ...
  adaptations: {
    // Don't load _any_ adaptations from these plugins:
    excludePlugins: [
      pluginX,
      pluginY,
    ],
    // Don't load these particular adaptations:
    exclude: [
      pluginA.adaptations.foo,
    ],
    // Load these non-plugin adaptations:
    include: [
      myAdaptation1,
      myAdaptation2,
    ],
  },
});
```


# Future concerns and thoughts


## Errors in adaptations

Adaptations can (and therefore, in the future, will) have errors in them. That will propagate upwards the component tree to the nearest ErrorBoundary.

This PR currently won't wrap adaptation logic in try/catch (`interceptProps`) or Provider components in local `ErrorBoundary`. Perhaps that would be good to do. That way, we could silently unfail adaptations, but leave the component be. The error can be sent to the error log.


## Link

The Link (which is turned into an adaptable component in this PR) has analytics logic hard-coded in it. Using the mechanisms in this PR, that could be replaced with an analytics plugin *adapt* Link and inject such logic that way.

We need a lot less coupling with esoteric logic in components when they are adaptable like this.


## Adapt-all-the-things

It's probably not a good idea to preemptively turn all components into adaptable ones, but rather on a by-need basis. Especially as it's not always obvious what _part_ of a component should be adaptable. As mentioned here, the `Link` is certainly useful. So are probably `Button`, `Avatar`, `EmptyState`, `SupportButton` and a few more. Apart from that, it would make sense for contributors to send PR's to turn components into adaptable ones when needed, as well as _expanding on what can be adapted_ (for already adaptable components) together with a brief rationale of _why_ this is needed.


## Example

This PR contains a commit in which the _GitHub actions_ plugin adapts Link. This is a dummy example, so THIS PR SHOULDN'T BE MERGED until that commit is removed. It's there to play around with for those who check out this branch locally. It currently appends "(this link goes to GitHub)" to each link's text, if the link goes to github.com. One can imagine instead rendering useful badges, hover tooltips or similar.

kuangp commented on packages/core-app-api/src/app/AppManager.tsx:
```diff
@@ -198,6 +202,15 @@ export class AppManager implements BackstageApp {
+      ...this.getPlugins().flatMap(plugin => plugin.extensions ?? []),
+      // App-provided extensions
+      ...this.extensions,
+    ];
```
Is there any precedence when there are conflicting global extensions? An app integrator specifying conflicting extensions for the same component is probably a bug but two different plugins may provide conflicting extensions of the same component

grantila replied:
I wouldn't necessarily consider multiple extensions to the same component _conflicting_. Instead, it's a feature that we can have multiple registered extensions of the same component. Extending components this was is really powerful and great, but can also be a can of worms. We _might_ end up with two extension of the same component _not playing well together_. I think we'll need to resolve that on a per-case basis.

In most cases, plugins (or the app itself) will extend components in a way that won't interfere with other extensions. Well have things like:
 * adding an action button
 * setting the title of something
 * rewriting the Link "to" (_only if it matches a certain pattern_, and otherwise fallthrough)

So, when multiple extensions to, say, the Link component are registered, the effective component becomes:

```tsx
<LinkProvider to={ /* ... */ }>
  <ExtensionA>
    <ExtensionB>
      <ExtensionC>
        <LinkComponent />
      </ExtensionC>
    </ExtensionB>
  </ExtensionA>
</LinkProvider>
```

If an intermediate component doesn't have any logic for this particular link (maybe the `to` doesn't match what the extension supports), the extension will just fallthrough with `return <Component />` which renders the next extension and eventually the inner component.

kuangp replied:
> We might end up with two extension of the same component not playing well together. I think we'll need to resolve that on a per-case basis.

Yeah that was my main concern (especially coming from different plugins who are unaware of other potential extensions) along with the ordering of plugins effecting the resulting extension outcome. 

This of course wouldn't be a concern under the opt-in model

kuangp commented on packages/core-plugin-api/src/extendable-components/createExtendableComponent.tsx:
```diff
@@ -0,0 +1,137 @@
+    );
+
+    return (
+      <componentRef.Provider
```
Would it make sense to also export the default `Provider` if someone wanted to restore the default behavior deeper within a nested `ProvideExtension` tree or would this be an anti-pattern? I'm not sure what this would mean for already intercepted props though ...

kuangp replied:
I like this approach 👍

kuangp commented on packages/core-plugin-api/src/extendable-components/types.ts:
```diff
@@ -0,0 +1,177 @@
+export type ComponentExtension<Props extends {}, Context> = {
+  ref: ExtendableComponentRef<Props, Context>;
+  spec: ExtendableComponentSpec<Props, Context>;
+  plugin?: BackstagePlugin<any, any>;
```
How do you envision this being used? eg. metadata to debug the source of plugin extensions?

grantila replied:
Yes and yes. Good digging.

This is currently not used, but I was thinking about e.g. error handling. Since extensions shouldn't be seen as (IMO right now) critical, but rather _nice_ improvements to existing components, it would be good if they wouldn't _fail_ the component if they threw an error e.g. Such an error could be caught and logged, and then knowing which plugin provided this extension would be useful. Especially in production builds, we could get useful messages. If we maybe also force extendable components to have id's, you could get:

`Component extension for 'link' provided by plugin 'github-actions' failed: Error: undefined is not a function`

Again, this is not implemented, but to tie a plugin to an extension was a way to _enable_ this in the future.

kuangp reviewed (commented):
This is pretty cool stuff and will be awesome for customization! 

My vote is for an opt-in model for plugin extensions. It's easier to be more intentional on what changes you want to pull in when you update plugins rather then be potentially surprised when a component changes and part of your app starts behaving differently (which is very easy to go unnoticed without thorough UI tests, also any new extensions of core-components could be considered a breaking change). It's the same workflow for when a plugin exports new components and its the responsibility of the integrator to "install" them if they are useful. I think plugin authors should scope extensions locally with  `ProvideExtension` where necessary  and also export `ComponentExtension`s to make them available.

grantila commented on packages/core-plugin-api/src/plugin/types.ts:
```diff
@@ -86,6 +91,7 @@ export type PluginConfig<
   id: string;
   apis?: Iterable<AnyApiFactory>;
   routes?: Routes;
+  extensions?: ComponentExtension<any, any>[];
```
Yes 👍

grantila commented:
Thanks @kuangp, really great feedback!

grantila commented:
Added required `id` to extendable components (as discussed in https://github.com/backstage/backstage/pull/11819#discussion_r888715834).

If we get something like https://github.com/backstage/backstage/pull/10260 merged, it could show a list of extensions and which one of them that are added to the app and which aren't.

Rugvip commented on packages/core-app-api/src/app/AppManager.tsx:
```diff
@@ -166,6 +168,7 @@ export class AppManager implements BackstageApp {
   private readonly icons: NonNullable<AppOptions['icons']>;
   private readonly plugins: Set<CompatiblePlugin>;
   private readonly components: AppComponents;
+  private readonly extensions: ComponentExtensions;
```
Would be nice to already now find a different name here tbh, perhaps to the concept as a whole. The "extension" name is currently occupied by the extensions that plugin export to make their content available to apps.

grantila replied:
Yeah I realised that way too late 🙈, that's an unfortunate naming collision.

Augmentation? Decoration? Addition? Customization?

Or, can we un-"extension" the current components, which are basically just components _bound_ to a plugin? Yes it'd be breaking, but could be smoothly done over time by allowing the old name too.

Rugvip replied:
We're likely about to have "extensions" become a thing in the backend system as well, with a somewhat similar role to the existing frontend extensions 😅.

Don't mean to bring the bike shed yet tbh, I think we can focus on the other parts first, just wanted to bring it up so that we can think about it a bit.

grantila replied:
I'ma keep bike shedding while we're on that topic. How about component "addon"?

grantila replied:
Yeah... `createComponentSwizzle()`, `swizzleComponent()`, `<ProvideSwizzle>`...

I don't think techdocs owns the 'addon' wording though, but maybe 'decoration' is the closest thing that somewhat accurately describes what this is.

grantila replied:
Just to be clear, what do the 'extension' word mean in `createComponentExtension`? What is _extending_ what there? Reading [this](https://backstage.io/docs/plugins/composability#extensions), it seems to basically be anything the plugin exports where the thing is _tied_ to the plugin.

This wording is a bit unfortunate, as it hijacks the word from where it _really_ would make sense, which is when _extending_ something. `createRoutableComponent` and `createComponent` instead of `*Extension`, or something like `exposeRoutableComponent` / `exposeComponent` would solve that. But yeah, that'd be breaking, eventually.

I'm leaning towards using _decoration_ here because of this, and piggy-back on the `.provide()` to get a uniform API.

Rugvip commented on packages/core-app-api/src/app/AppManager.tsx:
```diff
@@ -198,6 +202,15 @@ export class AppManager implements BackstageApp {
+  getExtensions(): ComponentExtensions {
+    return [
+      // Plugin-provided extensions
+      ...this.getPlugins().flatMap(plugin => plugin.extensions ?? []),
```
The ordering of the plugins list is deterministic, but not really from the users point of view. i.e. small changes to how the app is wired can have a large impact on the ordering of plugins. I think it's important that if we were to do anything like this, the ordering of the plugins has no affect on the rendered extensions.

grantila replied:
Both you and @kuangp prefer opt-in, and if we go that way, I think this won't be an issue. But you're right, if we keep this opt-out, it'll be quite confusing how multiple extension of the same component end up with a different result because of plugin loading order changes.

Rugvip commented on packages/core-plugin-api/src/extendable-components/ExtensionsProvider.tsx:
```diff
@@ -0,0 +1,162 @@
+  extensions: ComponentExtension<any, any>[] | ComponentExtension<any, any>;
+};
+
+type ExtensionMapKey = ExtendableComponentRef<any, any, any>;
```
The component ref objects aren't guaranteed to be unique of any given extendable component. You can and most likely will have duplicate installations of for example `@backstage/core-components`. I think the best approach is most likely to add a string ID for each component ref and use that as the key instead.

grantila replied:
Is that how api refs are handled internally? If so, yes it makes a lot of sense.

Rugvip replied:
Yep, they're keyed by the ID

Rugvip commented on packages/core-plugin-api/src/extendable-components/createExtendableComponent.tsx:
```diff
@@ -0,0 +1,137 @@
+    ? ({ ...config } as ExtendableComponentRef<Props, Context, RefInfoProps>)
+    : ({ ...config } as ExtendableComponentRef<Props, Context>);
+
+  const ctx = createContext<Context>(null as any);
```
```suggestion
  const ctx = createContext<Context | undefined>(undefined);
```

And add a check that it's available when consuming, for safety :grin:

Similar to the ref concern above, this context isn't guaranteed to be unique for each extendable component, would need to work around that somehow too.

Rugvip commented on packages/core-plugin-api/src/extendable-components/createExtendableComponent.tsx:
```diff
@@ -0,0 +1,137 @@
+            },
+          // Reduce using the real inner component as start value:
+          (() => {
+            // eslint-disable-next-line react-hooks/rules-of-hooks
```
I think you could clean this up if you defined these as functional components within the scope of the useMemo callback? Also an opportunity to give them a name for debuggability too.

Rugvip reviewed (commented):
Thank you! 👍 🎉 

I agree with @kuangp that opt-in is likely the way to go. It's simply too risky to have potential global app breakages whenever you bump a plugin. Upgrades are already hairy enough, we don't need to add more excitement into the mix. It should also help sort out any ordering issues one might have as well. That said we could also make it a difference between internal vs external plugins as well. So that open source plugins don't register extensions, but internal ones do.

Got one high level discussion that I want to dig into as well, as it affects pretty much all parts of this system:

What are the benefits of the double layered provider-based approach as opposed to an arguably simpler component decoration/middleware approach? i.e.

```tsx
extendComponent(linkComponentRef, {
  decorate(props, next) {
    if (!isLinkToPr(props.to))
      return next(props);

    const status = useGithubPrStatus(getPrFromLink(props.to));
    if (!status) {
      return next(props)
    }

    return <Badge color={prColorFromStatus(status)}>{next(props)}</Badge>
  },
})
```

or

```tsx
extendComponent(userCardComponentRef, {
  decorate(props, next) {
    if (!shouldHaveExtraActionButton(props)) {
      return next(props);
    }

    return next({
      ...props,
      actionButtons: [
        ...props.actionButtons,
        {
          text: 'My action',
          onClick: () => { /* ... */ },
        }
      ]
    })
  },
})
```

grantila commented:
Thanks for the feedback @Rugvip!

I think the main reasons for actually using _components_, rather middleware approach would be:
 * Middleware semantics are kind of broken similarly to callback based asynchrony (unlike promises). Missing to call `next()`, where would thrown errors end up?, calling `next()` but not returning but doing more things, maybe throwing, calling `next()` twice, etc. I've always found the code flow in callback based logic to easily become very iffy and hard to reason about, and hard to test. Unlike with promise/observable semantics, which React components are closer to.
 * Hook safety. An extension will likely use some hooks and then if/else return different pathways. If you chain multiple such extensions, the number of hooks (and the _same_ hooks) might not match every render, which they must. Letting each component be a pure regular component, makes this a lot more clear and easy to reason about. We can also write test helpers to easily test these extensions (by wrapping the props it'll get).
 * Mixing middleware `next` with components: `<Badge>{next()}</Badge>` is iffy.

If very often returning the same glue (when only changing context values) is too expressive and cumbersome:

```tsx
return <ComponentProvider value={...}>
  <Component />
</ComponentProvider>
```

we could certainly have a shortcut prop you can use, (called `Next`, say) but still being a real component:

```tsx
return <Next value={...}/>
```

Wouldn't that feel very reasonable? It's practically as clean, but won't cause confusion as to what "next" is.

But then, maybe `ComponentProvider` could be allowed to be childless to do the same thing? If we rename `ComponentProvider` to `Extend`, it can be both these?

```tsx
return <Extend value={...} /> // without children
// is a shortcut form of:
return <Extend value={...}><Component /></Extend>
```

grantila commented:
Playing around with this code, and thinking twice about that shortcut, I wonder if it would make more sense to only have `Component`.

 * If returned as is `return <Component />` then you don't alter the context. You can still wrap it in anything, like `<Tooltip>`.
 * If returned with a `value`: `<Component value={...} />` then you alter the context.

No need for an explicit provider, no shortcut syntax. We could call this `<Next />` and `<Next value={...} />` instead, but that name wouldn't be as indicative of that what you render is (most likely) the real inner component.

Rugvip commented:
To clarify, the intentions of the `next` function is that it's a render method, wrote it like that just to look more like middleware. Another way to write it that is a bit closer to your structure would be this:

```tsx
extendComponent(linkComponentRef, {
  decorate(props, Component) {
    if (!isLinkToPr(props.to))
      return <Component {...props} />

    const status = useGithubPrStatus(getPrFromLink(props.to));
    if (!status) {
      return <Component {...props} />
    }

    return (
      <Badge color={prColorFromStatus(status)}>
        <Component {...props} />
      </Badge>
    )
  },
})
```

or

```tsx
extendComponent(userCardComponentRef, {
  decorate(props, next) {
    if (!shouldHaveExtraActionButton(props)) {
      return <Component {...props} />
    }

    return <Component
      {...props}
      actionButtons={[
        ...props.actionButtons,
        {
          text: 'My action',
          onClick: () => { /* ... */ },
        }
      ]}
    />
  },
})
```

Either or really, don't have any strong opinion on which way. Important thing is that hooks work.

grantila commented:
Yeah, I think we're approaching this very similarly. Only having `Component` is extremely slick. Whether context providers are used or not can be entirely transparent to the API surface.

Rugvip commented on packages/core-components/src/components/Link/Link.tsx:
```diff
@@ -75,17 +79,37 @@ const getNodeText = (node: React.ReactNode): string => {
- */
-export const Link = React.forwardRef<any, LinkProps>(
-  ({ onClick, noTrack, ...props }, ref) => {
+export interface LinkContext {
+  to: string;
+  children: React.ReactNode;
+}
```
Wanna dig into this more as it was part of what I wanted to get at with the initial discussion. What are the benefits of this extra layer? More specifically the context type and the `value` prop. Could we instead use the existing props directly and only?

grantila replied:
The idea is that this extra layer is what the component _exports_ as actually changeable.

 * Props are what the user uses to the create component at _use-time_.
 * The context values are what _extensions_ use to _alter_ the component. It's kind of the internal _state_.

Consider `<UserCard userEntityRef={myRef} />` where the only prop here is `userEntityRef`, which you probably don't want to change (it wouldn't be _that_ useful). What this PR provides, is a way to change how that UserCard actually works, what it shows (e.g. action buttons, titles, avatars), etc.

So the UserCard parses the `userEntityRef`, get's the user info from the catalog, prepares data structures and then creates internal state out of this. Most components do this already, but they do it with `useState`. Then that _state_ is what is used to _render_ the output. If we instead formalise this state into one single context object and push that to the second (inner) _rendering component_, that's when we can let other intermediate components _alter_ the state and therefore change what the component does.

So, you can argue that this adds complexity or at least adds another layer, but you can also reason that it's actually just formalising the internal state components already have, into a _"publicly"_ interceptable state. You likely won't add complexity, just refactor and extract the preparation phase and rendering phase into two separate components.

And all parts of this;
 * the "setup" phase (called Provider here, maybe Setup is better)
 * the "render" phase (called Component here, maybe Render is better)
 * all extensions

all of those are pure React components with proper React lifecycle handling, so they can use hooks, asynchrony, etc.

In the Link example, it makes sense to let extensions change the `children` so they can decorate the UI, but in most components, this probably won't make much sense.

Rugvip replied:
I don't actually see how the context helps with your `userEntityRef` example, seems you could do similar things using just a middleware-like approach?

The one thing I do see is the potential information loss, for example if the context has `{x: number, y: number}` but the props is just `{sum: number}`, then you can't really reverse that operation. That might just be a bit of a straw man though tbh, because you could pass `x`, `y` through props. 

There's also the versioning and API stability aspect, but tbh I don't see a big benefit here either. The props are obviously already public API, and unless we remove the possibility to read props or override the render component then they're also both input and output types. Adding context on top of that just means we have a second type to manage, again both input and output. Even if that type is simpler, it's still extra.

so TL;DR: to me it seems like the context addition is only worthwhile if we keep the component props write-only, in which case it arguably simplifies the API surface.

grantila commented:
* Renamed extendable/extension adaptable/adaptation. Adaptation component `Provider` -> `Adaptation`.
* Both adaptable components and adaptations themselves can now be lazy-loaded (dynamic import)
* No forced opt-out, but either opt-in or opt-out
* Can create adaptable component refs separately from implementation (good for dynamically imported implementations)
* Adaptations are provided by plugins using `createComponentAdaptationExtension` (which creates an `Extension<T>`)

## Opt-in / opt-out

Now when spinning up the app, by default no adaptations are loaded. You can opt-in and load them individually. Plugins provide them using `createPlugin` and they are available (named) on the plugin object:

```tsx
createApp({
  // ...
  adaptations: [
    githubActionsPlugin.adaptations.link,
    // ... more adaptations
  ],
});
```

or you opt-out and load _all_ plugins' adaptations by default:

```tsx
createApp({
  // ...
  adaptations: {
    // Don't load _any_ adaptations from these plugins:
    excludePlugins: [
      pluginX,
      pluginY,
    ],
    // Don't load these particular adaptations:
    exclude: [
      pluginA.adaptations.foo,
    ],
    // Load these non-plugin adaptations:
    include: [
      myAdaptation1,
      myAdaptation2,
    ],
  },
});
```

## Separate component ref factory

When creating an adaptable component, you can create the component ref first:

```tsx
const myComponentRef = createAdaptableComponentRef({ id: 'my-component:v1' });
```

Then implement it:

```tsx
const MyComponent = implementAdaptableComponent(myComponentRef, {
  Provider() {
    // ...
  },
  Component() {
    // ...
  },
});
```

This can make sense if you want to structure a plugin that way.

## Lazy-loading

`createAdaptableComponent`, `createAdaptableForwardableComponent` (and `implementAdaptableComponent`) can now either take a `Provider` component _or_ an `asyncProvider` function. The same goes for the `Component` and `asyncComponent`. This allows for lazy-loading the component implementation.

```tsx
createAdaptableComponent<Props, Context>({
  id: 'MyComponent:v1',

  Provider: ({ props, Component }) => {
    return <Component value={...} />;
  },
  // or
  asyncProvider: () => import('./my-component').then(m => m.Provider),

  Component: ({ value }) => {
    // ...
  },
  // or
  asyncComponent: () => import('./my-component').then(m => m.Component),
});
```

It's still as typesafe.

For the adaptations, the same goes there:

```tsx
createComponentAdaptationExtension(fooComponentRef, {
  id: 'my-plugin-foo-adaptation',

  Adaptation() { /* ... */ },
  // or
  asyncAdaptation: () => import('./extensions/foo').then(m => m.Adaptation),
}),
```

### Lazy-loaded adaptations

When a component renders and has an adaptation provided that is lazy-loaded, by the nature of adaptations (they aren't supposed to be _necessary_, but entirely opt-in), the adaptation will load asynchronously, and the component will still render without the adaptation. When loading is complete, the adaptation is then applied. This means that there will be no _lag_ to wait for adaptations before rendering components.

grantila commented:
Updated PR message to reflect the latest changes

grantila commented:
Fixed the issue of having a versioned context keyed on component ref _objects_ (which could have multiple instances), into using the _id_ (string) instead.

As [discussed](https://github.com/backstage/backstage/pull/11819#discussion_r888712111) with @kuangp, being able to locally (in terms of React component tree) "unregister" previous adaptations would be good; Added `reset`, `excludePlugins`, `excludeComponents` and `exclude` to the `AdaptationProvider`:

```tsx
<AdaptationProvider
  reset                     // Resets _all_ previous adaptations downstream from here
  excludePlugins={[...]}    // Exclude (reset) all previously registered adaptations from these plugins
  excludeComponents={[...]} // Exclude all previously registered adaptations for this/these component ref(s)
  exclude={[...]}           // Exclude specifically these previously registered adaptations
  adaptations={...}         // Start using these adaptations
>
  <ChildrenGoHere />
</AdaptationProvider>
```

All props are optional, so you can e.g. reset (but not having to add new adaptations), or just add new (without resetting previous ones), or any combination.

It's also not an error to _exclude_ adaptations or plugins that actually weren't registered. That'll just be ignored.

kuangp commented on packages/core-plugin-api/src/adaptable-components/AdaptationProvider.tsx:
```diff
@@ -0,0 +1,260 @@
+            ({ adaptation }) =>
+              !adaptation.plugin || excludePluginsSet.has(adaptation.plugin),
+          )
+          .filter(({ adaptation }) => excludeSet.has(adaptation.key));
```
Am I misunderstanding here or should this condition be flipped? 
`.filter(({ adaptation }) => !excludeSet.has(adaptation.key));`
same with line 145 above?

grantila replied:
You're absolutely right, thanks!

All of this needs unit tests for sure, I've waited with that to get the API bounced here for discussion first though.

Rugvip commented on plugins/github-actions/src/plugin.ts:
```diff
@@ -23,10 +23,18 @@ import {
+  adaptations: {
+    link: createComponentAdaptationExtension(linkComponentRef, {
+      id: 'github-actions-link-adaptation',
+      asyncAdaptation: () => import('./extensions/link').then(m => m.Provider),
```
```suggestion
      asyncAdaptation: () => import('./adaptations/Link').then(m => m.Provider),
```

Wondering whether we should move the entire adaptation the separate file as the preferred pattern? Or do you think there's information here that's good to highlight?

Just for the purpose of keeping related information in the same place, no super strong opinion here. The dynamic import ofc gets a bit meh to define with an extra file but we can prolly handle that.

Rugvip commented on plugins/github-actions/src/plugin.ts:
```diff
@@ -23,10 +23,18 @@ import {
 export const githubActionsPlugin = createPlugin({
   id: 'github-actions',
+  adaptations: {
+    link: createComponentAdaptationExtension(linkComponentRef, {
```
Could call this `createComponentAdaptation` right?

Rugvip replied:
Ah right didn't realize this was actually an extension as well, should make sense if we structure it like that. I'm not entirely sure this should be an extension though, but perhaps something we can come back to after settling other pieces first.

grantila replied:
Yeah, the benefit of extensionalizing adaptations is the its logic can be wrapped in / tied to the plugin that provides it, so that errors thrown within an adaptation (which is "ok", or rather "handled" - that adaptation step will be dropped) can be bound to the plugin for ownership / debugging purposes.

Rugvip commented on plugins/github-actions/src/extensions/link.tsx:
```diff
@@ -0,0 +1,45 @@
+import { ComponentAdaptationOf } from '@backstage/core-plugin-api';
+import { linkComponentRef } from '@backstage/core-components';
+
+export const Provider: ComponentAdaptationOf<typeof linkComponentRef> = ({
```
```suggestion
export const Adaptation: ComponentAdaptationOf<typeof linkComponentRef> = ({
```

What do you think about something like that?

On a separate note we could do something like the ApiRefs do and make `typeof linkComponentRef.T` possible?

grantila replied:
The `T` of a component ref can refer to multiple things, the Props, Context, Component itself or the Adaptation of it. We could have `.TProps`, `.TContext`, `.TComponent`, `.TAdaptation` so that you could:

```tsx
typeof linkComponentRef.TAdaptation
// ===
ComponentAdaptationOf<typeof linkComponentRef>
```

I'm not sure which is nicest, I kind of like `YadaOf<typeof yada>` as a design pattern. It's very common (in fact, in many languages). The inverted opaque non-value `.T` is perhaps shorter in syntax, and you don't need to know what the `*Of<>` wrapper is called (and you don't need to import it), and that's nice. But it's also a little backwards, and you need to fake/null the property when constructing the object. Do we have a preferred way of dealing with this in Backstage?

Rugvip replied:
Yep no this makes if there are multiple different types to extract

Rugvip commented on packages/core-plugin-api/src/adaptable-components/types.ts:
```diff
@@ -0,0 +1,288 @@
+   * An opaque component representing the _inner_ component (or the next
+   * adaptation) with the default context value.
+   */
+  Component: ComponentType<OpaqueComponentProps<Context>>;
```
`ComponentType` is kinda meh because it adds `children` regardless of whether the component actually renders them or not. We tend to use something like `(props: TProps) => JSX.Element | null` instead.

grantila replied:
This is fixed in React 18 where `children` is no longer a default prop.

Rugvip replied:
Yep but not there yet. Feeling it's better to keep things aligned for now and then perhaps in the future we can migrate over to using the new `ComponentType`.

Rugvip commented on packages/core-plugin-api/src/adaptable-components/types.ts:
```diff
@@ -0,0 +1,288 @@
+import { BackstagePlugin } from '../plugin';
+
+/** @public */
+export type OpaqueComponentProps<Context extends {}> = { value: Context };
```
Please prefix type parameters with `T*`, https://github.com/backstage/backstage/blob/master/STYLE.md#naming

Rugvip replied:
Yep everything, the `T*` is there to make it easier to read the type definitions without wondering what's an arg and what isn't.

Rugvip reviewed (commented):
Alright, another round :grin:

I like where this is going but also getting worried it's sprinting ahead a bit too far, so I think it's best if we dial back a little bit to settle the core concepts first.

grantila commented on packages/core-app-api/src/app/types.ts:
```diff
@@ -218,6 +240,16 @@ export type AppOptions = {
+   * If provided as an array, then only these adaptations will be used.
+   *
+   * Oherwise provided as a {@link ComponentAdaptationOptions} object, in which
+   * case all the adaptations form plugins are included by default.
```
I figured both, defaulting to opt-in yes.

But _if you want_, you could instead opt-out and select adaptations or whole plugins to exclude adaptations from. Wouldn't this make sense?

grantila replied:
Sounds good, do we have an ADR or other precedent for naming features/functionality experimental?

grantila commented on packages/core-plugin-api/src/adaptable-components/AdaptationProvider.tsx:
```diff
@@ -0,0 +1,260 @@
+ * Prop types for the ApiProvider component.
+ * @public
+ */
+export type AdaptationProviderProps = {
```
Yeah, just `adaptation` and possibly a `reset` boolean I guess would make sense.. I might've jumped ahead a bit.

grantila commented:
# [Renamed, refactored] Adaptable components

The latest commits changes how these adaptable components are created in a big way, after discussions with the maintainers.

Instead of having adaptable components have props only read-only through the adaptation chains, and a separate type _Context_ being the values that can be (are allowed to be) changed, it's all props now.

An adaptable component, say `Link` has its props, same as before. Some of these props can be altered by adaptations. So, the interface for adaptations is the pair of `LinkProps` and a string union of which of the props are _adaptable_, e.g. `'to' | 'children'`.


## Abstract

We still have the ability for adaptable components to have an _outer_ and an _inner_ part, inbetween which adaptations will be injected.

The _outer_ part is called the `Prepare` component. The _inner_ is simply called `Component`. When this is rendered, the result will be:

```tsx
<Prepare>
  <Adaptation1>
    <Adaptation2>
      ...
      <Component>
```

Unless the adaptable component needs to _prepare_ props for adaptations, the `Prepare` can be ignored. This will be common.

Both of these are React components. For dynamic imports, you can instead provide `asyncPrepare` and `asyncComponent` when creating an adaptable component.

## API: Adaptable component

An example of the adaptable `Link` component:

```tsx
const adaptableLink = createAdaptableForwardableComponent<
  LinkProps,        // The adaptable Link has these props
  'to' | 'children' // And these props are adaptable, the rest not (readonly)
>({
  id: 'link:v1',
  Component: ({ props: { onClick, noTrack, ...props }, info: { ref } }) => {
    // ... actual <Link> code goes here.
    return <MuiLink {...someProps} />;
  },
});

export const linkComponentRef = adaptableLink.componentRef;
export const Link = adaptableLink.Component;
```

In this example, we skip providing a `Prepare` component, as we don't need to prepare any data for adapations, but it'd be a component side-by-side with `Component`.


## API: Adaptation

An adaptation is a component and/or a simple "props interceptor" (a sync function `props => adaptable props`).

The adaptation is exposed/exported when creating a plugin, and added to the `adaptations` property. An adaptation is created using the `componentRef` to refer to the adaptable component. This binds them together, and ensures type safety.

```tsx
// "My plugin" provides an adaptation to the "foo" component
// which may reside in another plugin

import { fooComponentRef } from '@acme/components';

export const myPlugin = createPlugin({
  id: 'my-plugin',
  adaptations: {
    link: createComponentAdaptationExtension(fooComponentRef, {
      id: 'my-plugin-foo-adaptation',
      asyncAdaptation: () => import(/*...*/),
      interceptProps: props => ({someProp: 'some-value'}),
    }),
  },
  // ...
});
```


### Props interceptor

When all you need is a synchronous function rewriting props, use a props interceptor.

```tsx
export const myPlugin = createPlugin({
  id: 'my-plugin',
  adaptations: {
    link: createComponentAdaptationExtension(fooComponentRef, {
      id: 'my-plugin-foo-adaptation',
      interceptProps: props => ({someProp: 'some-value'}), // <-- this
    }),
  },
  // ...
});
```

The props interceptor takes a props object as input, and returns an object with the rewritten props. The return type excludes the props that shouldn't be rewritten, only the adaptable props should be included.


### Component-based adaptation

As in the example above;

```tsx
export const myPlugin = createPlugin({
  id: 'my-plugin',
  adaptations: {
    link: createComponentAdaptationExtension(fooComponentRef, {
      id: 'my-plugin-foo-adaptation',
      asyncAdaptation: () =>
        import('./adaptations/foo.tsx').then(m => m.Adaptation), // <-- this
    }),
  },
  // ...
});
```

This creates a dynamically loaded adaptation, which is defined in `./adaptations/foo.tsx`. A component-based adaptation is referring to the `componentRef` of the adaptable component, to get the types induced for the adaptable props. Specifically if using the _type value_ `TAdaptation` which holds the type of an adaptation of the component.

```tsx
// foo.tsx

// The fooComponentRef encapsulates the type of the component, and the adaptable keys
import { fooComponentRef } from '@acme/components';

export const Adaptation: typeof fooComponentRef.TAdaptation = ({
  props,
  Component,
}) => {
  // ... code goes here, e.g.: ...

  const shouldAdapt = decide(props.something);

  if (!shouldAdapt)
    // Don't adapt, just forward
    return <Component />;

  // Change/set some prop(s)
  return <Component set={{ someProps: 'some-value' }} />;
};
```

grantila commented:
ping @backstage/maintainers, last push is a heavily simplified version, as discussed previously.

I can rebase and fix conflicts if you feel confident this would suffice for merging.

Rugvip commented:
@grantila thank you! gonna try to find some time to have a proper look at this. I think it's best if you could handle the conflicts either way tbh? Just to make sure we're looking at the most up to date version

hspens commented:
This sounds very useful! Has this been put on hold in favour of something else?

Rugvip commented:
Yep closing this as the same kind of capability has been implemented in #30750, thank you for kickstarting this a long while ago @grantila! 😁 The solution in #30750 drew from this design when it comes to decoupling input/output prop types, but didn't go quite as far when it comes to layered adaptations. It's still something that can be implemented using API context overrides, but we're leaving that out for now.
